src/cluster_slot.c

```diff
@@ -0,0 +1,244 @@
+ *     to endorse or promote products derived from this software without
+ *     specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "server.h"
+#include "cluster.h"
+#include "cluster_slot.h"
+
+#define DEFAULT_SLOT -1
+#define DEFAULT_STAT 0
+#define UNASSIGNED_SLOT 0
+#define ORDER_BY_KEY_COUNT 1
+#define ORDER_BY_INVALID -1
+
+/* -----------------------------------------------------------------------------
+ * Cluster slot public APIs.
+ * More to be added as part of cluster.c/h refactoring task. See issue #8948.
+ * -------------------------------------------------------------------------- */
+
+int getSlotOrReply(client *c, robj *o) {
+    long long slot;
+    if (getLongLongFromObject(o,&slot) != C_OK ||
+        slot < 0 || slot >= CLUSTER_SLOTS)
+    {
+        addReplyError(c,"Invalid or out of range slot");
+        return -1;
+    }
+    return (int) slot;
+}
+
+/* -----------------------------------------------------------------------------
+ * CLUSTER SLOT-STATS command
+ * -------------------------------------------------------------------------- */
+
+typedef struct sortedSlotStatEntry {
+    int slot;
+    uint64_t stat;
+} sortedSlotStatEntry;
+
+static bool doesSlotBelongToCurrentShard(int slot) {
+    clusterNode *n = server.cluster->myself;
+    return (server.cluster->slots[slot] == n) || (nodeIsSlave(n) && (server.cluster->slots[slot] == n->slaveof));
+}
+
+static void markAssignedSlots(unsigned char *slots) {
+    for (int slot = 0; slot < CLUSTER_SLOTS; slot++) {
+        if (doesSlotBelongToCurrentShard(slot)) slots[slot]++;
+    }
+}
+
+static int countAssignedSlotsFromSlotsArray(unsigned char *slots) {
+    int count = 0;
+    for (int slot = 0; slot < CLUSTER_SLOTS; slot++) {
+        if (slots[slot]) count++;
+    }
+    return count;
+}
+
+static int countAssignedSlotsFromCurrentShard() {
+    int count = 0;
+    clusterNode *n = server.cluster->myself;
+    if (nodeIsSlave(n)) {
+        count = n->slaveof->numslots;
+    } else {
+        count = server.cluster->myself->numslots;
+    }
+    return count;
+}
+
+static void checkSlotAssignment(unsigned char *slots, int start_slot, int end_slot) {
```

Comment:
Also, this function is structurally the same as markAssignedSlots but this one has more parameterized. I guess the other function is maybe named better, so maybe we rename this markAssignedSlots() and then just call this?

Reply:
Agreed on the large similarity between the two; 
1. `markAssignedSlots()`, and
2. `checkSlotAssignment()`

We can form a wrapper and implementation relationship between the two.
Wrapper being the generalized one, it will pass `0` and `CLUSTER_SLOTS-1` as the `start_slot` and `end_slot`.
```
/* Mark the slots array given a range of slots, if the slot belongs to the current shard.
 * The slots array is later used by addReplySlotStats(),
 * to count the total number of slots assigned to the current shard. */
static void markAssignedSlotsGivenSlotsRange(unsigned char *slots, int start_slot, int end_slot) {
    for (int slot = start_slot; slot <= end_slot; slot++) {
        if (doesSlotBelongToThisShard(slot)) slots[slot]++;
    }
}

static void markAssignedSlots(unsigned char *slots) {
    markAssignedSlotsGivenSlotsRange(slots, 0, CLUSTER_SLOTS-1);
}
```